File: php/src/Base/Models/Circle.php
<?php

namespace Shapes\Base\Models;

use Shapes\Base\Commons\{BaseShape, Dimensional, Displayable};

final class Circle extends BaseShape implements Dimensional {
    use Displayable;

    private readonly float $radius;
    private ?string $color;

    public function __construct(float $radius) {
        $dim1 = $radius > 0 ? $radius : 0;

        parent::__construct([
            ["radius" => max($radius, 0)],
            ["diameter" => max($radius, 0) * 2],
            ["curvature" => 1 / max($radius, 0)],
        ]);

        $this->radius = $dim1;
        $this->color = "None";
    }

    public function getColor(): string {
        return $this->color;
    }

    public function setColor(string $color): void {
        $this->color = $color;
    }

    #[\Override]
    public function printDescription(): void {
        echo "Circle radius: {$this->radius}" . PHP_EOL;
        echo "Circle dimensions: {$this->formatDimensions()}" . PHP_EOL;
        echo "Circle color: {$this->color}" . PHP_EOL;
        echo "Circle area: {$this->area()}" . PHP_EOL;
        echo "Circle perimeter: {$this->perimeter()}" . PHP_EOL;
        echo "Circle is displayable: {$this->shouldDisplay()} " . PHP_EOL;
        echo "Circle is a shape: " . self::validateShape() . PHP_EOL;
    }

    #[\Override]
    public function area(): float {
        return M_PI * $this->radius ** 2;
    }

    #[\Override]
    public function perimeter(): float {
        return 2.0 * M_PI * $this->radius;
    }

    private function formatDimensions(): string {
        $parts = (array) array_map(
            static fn ($dimension) => implode(
                " | ",
                array_map(
                    static fn ($key, $value) => strtoupper($key) . ": " . number_format($value, 3),
                    array_keys($dimension),
                    $dimension
                )
            ),
            $this->getDimensions()
        );

        return implode(" | ", $parts);
    }
}

File: rust/src/base/models/circle.rs
use std::collections::HashMap;
use std::f64::consts::PI;

use crate::base::commons::{BaseShape, Dimensional, Displayable};

pub struct Circle {
    pub radius: f64,
    dimensions: Vec<HashMap<&'static str, f64>>,
    color: &'static str,
}

impl Circle {
    pub fn new(radius: f64) -> Self {
        let dim_one: f64 = if radius < 0.0 { 1.0 } else { radius };

        Self {
            radius: dim_one,
            dimensions: vec![
                HashMap::from([("radius", radius.max(1.0))]),
                HashMap::from([("diameter", radius.max(1.0) * 2.0)]),
                HashMap::from([("curvature", 1.0 / radius.max(1.0))]),
            ],
            color: "None",
        }
    }

    pub fn get_color(&self) -> &str {
        self.color
    }

    pub fn set_color(&mut self, color: &'static str) {
        self.color = color;
    }

    fn format_dimensions(&self) -> String {
        let parts: Vec<String> = self
            .dimensions
            .iter()
            .flat_map(|dimension| {
                dimension
                    .iter()
                    .map(|(key, value)| format!("{}: {:.3}", key.to_uppercase(), value))
            })
            .collect::<Vec<_>>();

        parts.join(" | ")
    }
}

impl BaseShape for Circle {
    fn get_dimensions(&self) -> &Vec<HashMap<&'static str, f64>> {
        &self.dimensions
    }

    fn print_description(&self) {
        println!("Circle radius: {}", self.radius);
        println!("Circle dimensions: {}", self.format_dimensions());
        println!("Circle color: {}", self.get_color());
        println!("Circle area: {}", self.area());
        println!("Circle perimeter: {}", self.perimeter());
        println!("Circle is displayable: {}", self.should_display());
        println!("Circle is a shape: {}", Circle::validate_shape());
    }
}

impl Dimensional for Circle {
    fn area(&self) -> f64 {
        PI * self.radius.powi(2)
    }

    fn perimeter(&self) -> f64 {
        2.0 * PI * self.radius
    }
}

impl Displayable for Circle {}
